drivers/led_strip/ws2812_gpio.c

```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/led_strip.h>
+
+#include <string.h>
+
+#define LOG_LEVEL CONFIG_LED_STRIP_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(ws2812_gpio);
+
+#include <zephyr.h>
+#include <soc.h>
+#include <drivers/gpio.h>
+#include <device.h>
+#include <drivers/clock_control.h>
+
+struct ws2812_gpio_data {
+	struct device *gpio;
+	struct device *clk;
+};
+
+struct ws2812_gpio_cfg {
+	u8_t pin;
+	bool has_white;
+};
+
+static struct ws2812_gpio_data *dev_data(struct device *dev)
+{
+	return dev->driver_data;
+}
+
+static const struct ws2812_gpio_cfg *dev_cfg(struct device *dev)
+{
+	return dev->config->config_info;
+}
+
+/*
+ * This is hard-coded to nRF51 in two ways:
+ *
+ * 1. The assembly delays T1H, T0H, TxL
+ * 2. GPIO set/clear
+ */
+
+/*
+ * T1H: 1 bit high pulse delay: 12 cycles == .75 usec
+ * T0H: 0 bit high pulse delay: 4 cycles == .25 usec
+ * TxL: inter-bit low pulse delay: 8 cycles == .5 usec
+ *
+ * We can't use k_busy_wait() here: its argument is in microseconds,
+ * and we need roughly .05 microsecond resolution.
+ */
+#define DELAY_T1H "nop\nnop\nnop\nnop\nnop\nnop\nnop\nnop\nnop\nnop\nnop\nnop\n"
+#define DELAY_T0H "nop\nnop\nnop\nnop\n"
+#define DELAY_TxL "nop\nnop\nnop\nnop\nnop\nnop\nnop\nnop\n"
+
+/*
+ * GPIO set/clear (these make assumptions about assembly details
+ * below).
+ *
+ * This uses OUTCLR == OUTSET+4.
+ *
+ * We should be able to make this portable using the results of
+ * https://github.com/zephyrproject-rtos/zephyr/issues/11917.
+ *
+ * We already have the GPIO device stashed in ws2812_gpio_data, so
+ * this driver can be used as a test case for the optimized API.
+ *
+ * Per Arm docs, both Rd and Rn must be r0-r7, so we use the "l"
+ * constraint in the below assembly.
+ */
+#define SET_HIGH "str %[p], [%[r], #0]\n" /* OUTSET = BIT(LED_PIN) */
+#define SET_LOW "str %[p], [%[r], #4]\n"  /* OUTCLR = BIT(LED_PIN) */
+
+/* Send out a 1 bit's pulse */
+#define ONE_BIT(base, pin) do {			\
+	__asm volatile (SET_HIGH			\
+			DELAY_T1H			\
+			SET_LOW			\
+			DELAY_TxL			\
+			::				\
+			[r] "l" (base),		\
+			[p] "l" (pin)); } while (0)
+
+/* Send out a 0 bit's pulse */
+#define ZERO_BIT(base, pin) do {			\
+	__asm volatile (SET_HIGH			\
+			DELAY_T0H			\
+			SET_LOW			\
+			DELAY_TxL			\
+			::				\
+			[r] "l" (base),		\
+			[p] "l" (pin)); } while (0)
+
+static int send_buf(struct device *dev, u8_t *buf, size_t len)
+{
+	volatile u32_t *base = (u32_t *)&NRF_GPIO->OUTSET;
+	const u32_t val = BIT(dev_cfg(dev)->pin);
+	struct device *clk = dev_data(dev)->clk;
+	unsigned int key;
+	int rc;
+
+	rc = clock_control_on(clk, NULL);
```

Comment:
Looks like some clock control related changes were lost in the rebase:

https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/led_strip/ws2812b_sw.c#L33

I see the DNM, but can you please fix that up here?